pkg/hostedMasterApi/v20170831/types.go

```diff
@@ -0,0 +1,224 @@
+type PublicKey struct {
+	KeyData string `json:"keyData"`
+}
+
+// WindowsProfile represents the Windows configuration passed to the cluster
+type WindowsProfile struct {
+	AdminUsername string `json:"adminUsername,omitempty" validate:"required"`
+	AdminPassword string `json:"adminPassword,omitempty" validate:"required"`
+}
+
+// ProvisioningState represents the current state of container service resource.
+type ProvisioningState string
+
+const (
+	// Creating means ContainerService resource is being created.
+	Creating ProvisioningState = "Creating"
+	// Updating means an existing ContainerService resource is being updated
+	Updating ProvisioningState = "Updating"
+	// Failed means resource is in failed state
+	Failed ProvisioningState = "Failed"
+	// Succeeded means resource created succeeded during last create/update
+	Succeeded ProvisioningState = "Succeeded"
+	// Deleting means resource is in the process of being deleted
+	Deleting ProvisioningState = "Deleting"
+	// Migrating means resource is being migrated from one subscription or
+	// resource group to another
+	Migrating ProvisioningState = "Migrating"
+)
+
+// NetworkProfile represents the definition of network profile for the cluster
+type NetworkProfile struct {
+	PodCidr          string `json:"podCidr,omitempty"`
+	ServiceCidr      string `json:"serviceCidr,omitempty"`
+	KubeDNSServiceIP string `json:"kubeDNSServiceIP,omitempty"`
+}
+
+// JumpboxProfile represents the definition of bastion machine to access the agent nodes
+type JumpboxProfile struct {
+	PublicIPAddressID string `json:"publicIPAddressID,omitempty"`
+	PublicIPAddress   string `json:"publicIPAddress,omitempty"`
+}
+
+// AgentPoolProfile represents configuration of VMs running agent
+// daemons that register with the master and offer resources to
+// host applications in containers.
+type AgentPoolProfile struct {
+	Name           string `json:"name" validate:"required"`
+	Count          int    `json:"count" validate:"required,min=1,max=100"`
+	VMSize         string `json:"vmSize" validate:"required"`
+	OSDiskSizeGB   int    `json:"osDiskSizeGB,omitempty" validate:"min=0,max=1023"`
+	StorageProfile string `json:"storageProfile" validate:"eq=StorageAccount|eq=ManagedDisks|len=0"`
+	VnetSubnetID   string `json:"vnetSubnetID,omitempty"`
+
+	// OSType is the operating system type for agents
+	// Set as nullable to support backward compat because
+	// this property was added later.
+	// If the value is null or not set, it defaulted to Linux.
+	OSType OSType `json:"osType,omitempty"`
+
+	// subnet is internal
+	subnet string
+}
+
+// AccessProfile represents role name and kubeconfig
+type AccessProfile struct {
+	RoleName   string `json:"roleName"`
+	KubeConfig string `json:"kubeConfig"`
+}
+
+// UnmarshalJSON unmarshal json using the default behavior
+// And do fields manipulation, such as populating default value
+func (a *AgentPoolProfile) UnmarshalJSON(b []byte) error {
```

Comment:
Check other examples of how validation is being done for API versions for container service. This is doing the same. At this time I am not planning to do validation in different ways for different API versions or resources types.